lib/web_ui/dev/watcher.dart

```diff
@@ -3,20 +3,85 @@
 // found in the LICENSE file.
 
 import 'dart:async';
+import 'dart:io' as io;
 
 import 'package:path/path.dart' as path;
 import 'package:watcher/watcher.dart';
 
+import 'utils.dart';
+
 enum PipelineStatus {
+  /// The pipeline has not started yet.
+  ///
+  /// This is the initial state of the pipeline.
   idle,
+
+  /// The pipeline is running build steps.
   started,
+
+  /// The pipeline is stopping.
   stopping,
-  stopped,
+
+  /// The pipeline is not running anything because it has been interrupted.
+  interrupted,
+
+  /// The pipeline is not running anything because it encountered an error.
   error,
+
+  /// The pipeline is not running anything because it finished all build steps successfully.
   done,
 }
 
-typedef PipelineStep = Future<void> Function();
+/// A step in the build pipeline.
+abstract class PipelineStep {
+  /// The name of this step.
+  ///
+  /// This value appears in logs, so it should be descriptive and human-readable.
+  String get name;
+
+  /// Whether it is safe to interrupt this step while it's running.
+  bool get isSafeToInterrupt;
+
+  /// Runs this step.
+  ///
+  /// The returned future is completed when the step is finished. The future
+  /// completes with an error if the step failed.
+  Future<void> run();
+
+  /// Interrupts this step, if it's already running.
+  ///
+  /// [Pipeline] only calls this if [isSafeToInterrupt] returns true.
+  Future<void> interrupt();
+}
+
+/// A helper class for implementing [PipelineStep] in terms of a process.
+abstract class ProcessStep implements PipelineStep {
+  ProcessManager? _process;
+  bool _isInterrupted = false;
+
+  /// Starts and returns the process that implements the logic of this pipeline
+  /// step.
+  Future<ProcessManager> createProcess();
+
+  @override
+  Future<void> interrupt() async {
+    _isInterrupted = true;
+    _process?.kill();
+  }
+
+  @override
+  Future<void> run() async {
+    final ProcessManager process = await createProcess();
+
+    if (_isInterrupted) {
+      return;
+    }
```

Comment:
There is a potential race condition here that could cause the process to not be killed.

1. Begin `run()`
2. Begin `createProcess()`
3. Call `interrupt()` while `createProcess()` is still running asynchronously.
4. `_process?.kill();` does nothing because `_process` is still null.
5. End `createProcess()`
6. `_isInterrupted` is true and we return from `run()`.
7. End `run()`

We end up in a situation where we think we interrupted the process, but we actually didn't.

Reply:
LOL, thanks for the catch! I was specifically trying to avoid this situation, which is why I assign `_process = process` _after_ the `if` block to avoid the race with `interrupt()`, but then I forgot to actually kill the process inside the if 🤦‍♂️ 

Done.